File: market/replay/replay_test.go
package replay

import (
	"errors"
	"testing"
	"time"

	"github.com/thaodangspace/tidekeepers-server/market/basket"
	"github.com/thaodangspace/tidekeepers-server/market/sector"
)

func TestVerifyBasketAndBenchmark(t *testing.T) {
	t.Parallel()

	basketInput := basket.CalculationInput{
		Mapping: basket.MappingVersion{ID: "mapping", Key: "crest", Sector: sector.Crest, MinimumCoveredWeight: 100_000_000, NormalizationCap: 2_000_000, ExpectedTurbulence: basket.ExpectedTurbulencePolicy{Type: basket.TurbulencePolicyStaticContentValue, StaticValue: 1, Floor: 1}, Components: []basket.MappingComponent{{MarketAssetID: "a", TargetWeight: 50_000_000, Enabled: true}, {MarketAssetID: "b", TargetWeight: 50_000_000, Enabled: true}}},
		Window:  basket.CalculationWindow{DailyTideID: "tide", ProviderKey: "fixture", Start: time.Date(2026, time.July, 30, 0, 0, 0, 0, time.UTC), End: time.Date(2026, time.July, 31, 0, 0, 0, 0, time.UTC)},
	}
	basketExpected, err := (basket.Calculator{}).Calculate(basketInput)
	if err != nil {
		t.Fatalf("Calculate basket: %v", err)
	}
	if err := VerifyBasket(basketInput, basketExpected); err != nil {
		t.Fatalf("VerifyBasket: %v", err)
	}
	basketExpected.Status = basket.MetricStatusReady
	if err := VerifyBasket(basketInput, basketExpected); !errors.Is(err, ErrMismatch) {
		t.Fatalf("VerifyBasket mismatch = %v", err)
	}

	first, second := int64(-100), int64(100)
	benchmarkInput := sector.BenchmarkInput{DailyTideID: "tide", Definition: sector.Definition{ID: "crest-v2", Sector: sector.Crest, BenchmarkMethod: sector.EqualWeight, MinimumEligibleBaskets: 2, RelativeScale: 500000, RelativeBlendWeight: 700000, RankBlendWeight: 300000, ScoreCap: 1000000}, Baskets: []sector.BenchmarkBasket{{BasketMetricID: "metric-a", BasketMappingVersionID: "map-a", Sector: sector.Crest, BenchmarkEligible: true, MetricStatus: "READY", NormalizedPerformance: &first}, {BasketMetricID: "metric-b", BasketMappingVersionID: "map-b", Sector: sector.Crest, BenchmarkEligible: true, MetricStatus: "READY", NormalizedPerformance: &second}}}
	benchmarkExpected, err := (sector.BenchmarkCalculator{}).Calculate(benchmarkInput)
	if err != nil {
		t.Fatalf("Calculate benchmark: %v", err)
	}
	if err := VerifyBenchmark(benchmarkInput, benchmarkExpected); err != nil {
		t.Fatalf("VerifyBenchmark: %v", err)
	}
	benchmarkExpected.EligibleBasketCount++
	if err := VerifyBenchmark(benchmarkInput, benchmarkExpected); !errors.Is(err, ErrMismatch) {
		t.Fatalf("VerifyBenchmark mismatch = %v", err)
	}
}

File: server/tests/replay_test.ts
import { assertEquals, assertRejects } from "@std/assert";
import { type CalculationInput, Calculator } from "../src/market/basket.ts";
import {
  BenchmarkCalculator,
  type BenchmarkInput,
  EqualWeight,
} from "../src/market/sector.ts";
import {
  replayMismatchMessage,
  verifyBasket,
  verifyBenchmark,
} from "../src/market/replay.ts";

Deno.test("replay: verifies basket and benchmark outputs exactly", async () => {
  const basketInput: CalculationInput = {
    mapping: {
      id: "mapping",
      key: "crest",
      sector: "CREST",
      minimumCoveredWeight: 100_000_000,
      normalizationCap: 2_000_000,
      benchmarkEligible: true,
      expectedTurbulence: {
        id: "turbulence",
        type: "STATIC_CONTENT_VALUE",
        staticValue: 1,
        floor: 1,
      },
      components: [
        {
          marketAssetId: "a",
          targetWeight: 50_000_000,
          enabled: true,
          sequence: 0,
        },
        {
          marketAssetId: "b",
          targetWeight: 50_000_000,
          enabled: true,
          sequence: 1,
        },
      ],
    },
    window: {
      id: "window",
      dailyTideId: "tide",
      providerKey: "fixture",
      start: "2026-07-30T00:00:00Z",
      end: "2026-07-31T00:00:00Z",
      toleranceMs: 0,
    },
    observations: [
      {
        id: "a-open",
        providerKey: "fixture",
        marketAssetId: "a",
        observedAt: "2026-07-30T00:00:00Z",
        priceUnits: 100_000_000,
        status: "VALID",
      },
      {
        id: "a-close",
        providerKey: "fixture",
        marketAssetId: "a",
        observedAt: "2026-07-31T00:00:00Z",
        priceUnits: 110_000_000,
        status: "VALID",
      },
      {
        id: "b-open",
        providerKey: "fixture",
        marketAssetId: "b",
        observedAt: "2026-07-30T00:00:00Z",
        priceUnits: 100_000_000,
        status: "VALID",
      },
      {
        id: "b-close",
        providerKey: "fixture",
        marketAssetId: "b",
        observedAt: "2026-07-31T00:00:00Z",
        priceUnits: 90_000_000,
        status: "VALID",
      },
    ],
  };
  const basketExpected = await new Calculator().calculate(basketInput);
  await verifyBasket(basketInput, basketExpected);
  basketExpected.rawReturn = 123;
  await assertRejects(
    () => verifyBasket(basketInput, basketExpected),
    Error,
    replayMismatchMessage,
  );

  const benchmarkInput: BenchmarkInput = {
    dailyTideId: "tide",
    definition: {
      id: "crest-v2",
      sector: "CREST",
      benchmarkMethod: EqualWeight,
      minimumEligibleBaskets: 2,
      relativeScale: 500_000,
      relativeBlendWeight: 700_000,
      rankBlendWeight: 300_000,
      scoreCap: 1_000_000,
    },
    baskets: [
      {
        basketMetricId: "metric-a",
        basketMappingVersionId: "map-a",
        sector: "CREST",
        benchmarkEligible: true,
        metricStatus: "READY",
        normalizedPerformance: -100,
      },
      {
        basketMetricId: "metric-b",
        basketMappingVersionId: "map-b",
        sector: "CREST",
        benchmarkEligible: true,
        metricStatus: "READY",
        normalizedPerformance: 100,
      },
    ],
  };
  const benchmarkExpected = await new BenchmarkCalculator().calculate(
    benchmarkInput,
  );
  await verifyBenchmark(benchmarkInput, benchmarkExpected);
  benchmarkExpected.eligibleBasketCount++;
  await assertRejects(
    () => verifyBenchmark(benchmarkInput, benchmarkExpected),
    Error,
    replayMismatchMessage,
  );
  assertEquals(benchmarkExpected.eligibleBasketCount, 3);
});
